# Computational Physics U24568
## Lecture 2a - Advanced Numpy and Data Manipulation

Today's lecture will focus on some more features in numpy exploring some more tools for manipulating data and quickly accessing things that are desired.

Numpy features that we will cover today:

* Boolean indexing, `logical_and` etc.
* Built in methods of the `numpy.array` class, including `.any`, `.all`. `.where`.
* Transpose and reshaping arrays.
* The numpy `matrix` class (which is a sub-class of `numpy.array`)
* How to create tables of data, using the pandas library, to efficiently hold and access datasets in memory
 * How to write these datasets to file.
 * How to read such datasets in from file.
 * How to load supplied datasets into this pandas format.
 * How to efficiently manipulate these data arrays to quickly access, or plot, the data you are interested in.

In [22]:
# We'll almost always import numpy as np to save some typing
import numpy as np

## Review - Basic Numpy

We'll start by reviewing some numpy basics. You can create a numpy array by
 1. Using `np.array` on a list of numbers.
 1. Using `np.arange` or `np.linspace` to create an equally-spaced array.
 1. Using `np.zeros` or `np.ones` to create an array of the same number.

**IMPORTANT**: It's best to vectorize your operations on numpy arrays. That means using numpy functions (and built-in math function like `+`, `-`, `*`, `/`, `**`, etc.) on the *entire array*. This is both clearer and faster than writing a loop over the indexes. Its best to try to do things only in numpy first if you can.

In [26]:
# Convert a list to a numpy array
a_list = [1,2,3,4,5,6,7,8,9,10]
a_numpy_array = np.array(a_list)
print(a_numpy_array)

[ 1  2  3  4  5  6  7  8  9 10]


In [24]:
# Create a range of values, just like the range builtin
x_vals = np.arange(0, 2*np.pi, np.pi/6)

# Calculate sin over the entire array
sin_x = np.sin(x_vals)
print(sin_x)

[ 0.00000000e+00  5.00000000e-01  8.66025404e-01  1.00000000e+00
  8.66025404e-01  5.00000000e-01  1.22464680e-16 -5.00000000e-01
 -8.66025404e-01 -1.00000000e+00 -8.66025404e-01 -5.00000000e-01]


In [13]:
# Test the trigonometric identity
np.cos(x_vals)**2 + np.sin(x_vals)**2

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

It is possible to write this as a loop, as done below. But it's more lines, easier to make a mistake, and slower when you have millions of numbers.

In [17]:
result = []
for i in range(len(x_vals)):
    x = x_vals[i]
    result.append(np.cos(x)**2 + np.sin(x)**2)

print(result)

[1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]


In [32]:
# Nine equally-spaced values from -1 to 1
print(np.linspace(-1, 1, 9))

# Note that printing the numpy array looks different from
# just outputting it (as the last line of the cell)
np.linspace(-1, 1, 9)

[-1.   -0.75 -0.5  -0.25  0.    0.25  0.5   0.75  1.  ]


array([-1.  , -0.75, -0.5 , -0.25,  0.  ,  0.25,  0.5 ,  0.75,  1.  ])

## Review - Slicing Arrays

We'll review how to select portions of the data using *slicing*. Let's start by defining an array to work with that has floating point numbers of the form 'row.column', so it's easy to see if we've got the right data.

In [34]:
arr10 = np.array([[(ii+0.1*jj) for jj in range(10)] for ii in range(10)])
print(arr10)

[[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9]
 [1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9]
 [2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9]
 [3.  3.1 3.2 3.3 3.4 3.5 3.6 3.7 3.8 3.9]
 [4.  4.1 4.2 4.3 4.4 4.5 4.6 4.7 4.8 4.9]
 [5.  5.1 5.2 5.3 5.4 5.5 5.6 5.7 5.8 5.9]
 [6.  6.1 6.2 6.3 6.4 6.5 6.6 6.7 6.8 6.9]
 [7.  7.1 7.2 7.3 7.4 7.5 7.6 7.7 7.8 7.9]
 [8.  8.1 8.2 8.3 8.4 8.5 8.6 8.7 8.8 8.9]
 [9.  9.1 9.2 9.3 9.4 9.5 9.6 9.7 9.8 9.9]]


Selecting one entry in the array uses the format `arr[row, col]`. Row 0, column 2 in the matrix above would be `arr10[0, 2]`.

In [36]:
arr10[0,2]

0.2

We can use the same slicing syntax used for lists on arrays as well. We just need to provide a slice for the row and the column. Here are some examples of 1D slices:

 1. `arr[1]` element 1 of arr
 1. `arr[1:4]` elements 1 up to 4 (but not including 4)
 1. `arr[1:]` element 1 to the end
 1. `arr[1:7:2]` element 1, 3, 5, up to 7
 1. `arr[1::2]` element 1, 3, 5, to the end
 1. `arr[:]` all of arr

 We can use any of these in the row and the column entry.

In [37]:
# Slicing - All rows, column 1
arr10[:,1]

array([0.1, 1.1, 2.1, 3.1, 4.1, 5.1, 6.1, 7.1, 8.1, 9.1])

In [38]:
# Slicing - Row 2, all columns
arr10[2,:]

array([2. , 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9])

In [39]:
# Slicing - Row 2, columns 1 through 7, skipping by 2
arr10[2,1:8:2]

array([2.1, 2.3, 2.5, 2.7])

In [40]:
# Slicing - Last row
arr10[-1,:]

array([9. , 9.1, 9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9])

We can also assign values by putting the slice before an equals sign. Using `:` we can do it to a whole column.

In [44]:
test_arr = arr10.copy()
test_arr[:,5] = 0.
print(test_arr)

[[0.  0.1 0.2 0.3 0.4 0.  0.6 0.7 0.8 0.9]
 [1.  1.1 1.2 1.3 1.4 0.  1.6 1.7 1.8 1.9]
 [2.  2.1 2.2 2.3 2.4 0.  2.6 2.7 2.8 2.9]
 [3.  3.1 3.2 3.3 3.4 0.  3.6 3.7 3.8 3.9]
 [4.  4.1 4.2 4.3 4.4 0.  4.6 4.7 4.8 4.9]
 [5.  5.1 5.2 5.3 5.4 0.  5.6 5.7 5.8 5.9]
 [6.  6.1 6.2 6.3 6.4 0.  6.6 6.7 6.8 6.9]
 [7.  7.1 7.2 7.3 7.4 0.  7.6 7.7 7.8 7.9]
 [8.  8.1 8.2 8.3 8.4 0.  8.6 8.7 8.8 8.9]
 [9.  9.1 9.2 9.3 9.4 0.  9.6 9.7 9.8 9.9]]


## New content - Boolean indexing

Let's start with boolean indexing on a numpy array. What does this mean?

Well let's sat that you have a set of numbers and you want to identify all those numbers whose value is larger than 2. You could write a for loop (which is often quite slow to run in python), or you could use "boolean indexing". This goes something like:

In [ ]:
import numpy as np
# Here's a for loop showing what we're doing
some_numbers = [1,10,2,20,3,30,4,40,5,50,6,60]
large_nums = []

for num in some_numbers:
    if num > 4:
        large_nums.append(num)

# But we could also do this as using boolean indexing:

some_numbers = np.array(some_numbers)
large_nums = some_numbers[some_numbers > 4]
print(large_nums)


Let's try and break down what this is doing. It only works with numpy arrays so if you have a list, first convert it to a numpy array by doing `some_numbers = np.array(some_numbers)`.

Then we do `some_numbers > 4`. This creates a boolean array (an array whose entries are either `True` or `False`).

We then use this boolean array to index `some_numbers`. It will extract all entries for which the boolean array is True. The boolean array must be the same length as the array being indexed.

So for example

`np.array([1,4,9,16])[np.array([True,False,False,True])]`

will return the first and fourth entries of this array (1 and 16) in a new array.

In [ ]:
np.array([1,4,9,16])[np.array([True,False,False,True])]

We can split this into two lines if we also want to store the Boolean array.

In [ ]:
some_numbers = np.array([1,10,2,20,3,30,4,40,5,50,6,60])
logic_array = some_numbers > 4
print(logic_array)
large_nums = some_numbers[logic_array]
print(large_nums)

**EXERCISE**

Let's practice this.

* Use boolean indexing to extract all values in some_numbers where the number is smaller than 15
* Use boolean indexing to extract all values in some_numbers where $\mathrm{num}^2 - 15$ is larger than 100

In [ ]:
some_numbers = np.array([1,10,2,20,3,30,4,40,5,50,6,60])

# COMPLETE BELOW



How is this better than using a `for` loop? It is significantly faster, so if you are dealing with very large arrays, you might want this. But another huge advantage to this comes when you're dealing with multiple arrays of related data. Let's illustrate this using a pandas DataFrame.

In [ ]:
import numpy as np
import pandas as pd
def func_makedata(a):
    x1 = a**2
    y1 = np.cos(a)
    z1 = 3*a**2 
    return x1, y1, z1

aa = np.linspace(0.,10.,50)
x1, y1, z1 = func_makedata(aa)

data_dict = {}
data_dict['aa'] = aa
data_dict['x1'] = x1
data_dict['y1'] = y1
data_dict['z1'] = z1
pd_dataframe = pd.DataFrame(data_dict)
pd_dataframe

This data structure contains 4 arrays of data, but these are all related to each other (in particular x1, y1 and z1 are all functions of the first array, aa). So how could we extract all values of x1 for which y1 is larger than 0.5? This could be done with a `for` loop, but the following approach is *significantly* faster and makes the code much more compact.

In [ ]:
# Create the boolean array
large_y1_logic = pd_dataframe['y1'] > 0.5
# Access corresponding values of x1
x1_values = pd_dataframe['x1'][large_y1_logic]
# Match this against the table above .. pandas also retains the indexes!
print(x1_values)

# Similarly we can quickly obtain the corresponding values of aa or z1
aa_values = pd_dataframe['aa'][large_y1_logic]
z1_values = pd_dataframe['z1'][large_y1_logic]

With these tools you **never need to use Excel again**. You now have access to a much more powerful, faster, and more flexible method to slice and dice data and to graphically display the output however you want! We'll learn more about this below.

**EXERCISES**

Let's try a few more exercises on our `pd_dataframe` dataset:

* Extract all values of `z1` for which the corresponding value of $\mathrm{x1} - 15$ is larger than 20.
* Extract all values of `x1` for which the corresponding value of $\mathrm{aa} * (\mathrm{x1} - 50) * \mathrm{y1}$ is larger than 20

Let's add one more layer of complexity to this. What if we want to access all values of `z1` for which $\mathrm{y1} > 0.5$ *and* $\mathrm{x1} < 10$. One could do some sort of nested approach (which actually might be the most computationally efficient way of doing this, but isn't the most elegant):

In [ ]:
y1_logic = pd_dataframe['y1'] > 0.5
reduced_x1 = pd_dataframe['x1'][y1_logic]
reduced_z1 = pd_dataframe['z1'][y1_logic]

final_z1_values = reduced_z1[reduced_x1 < 10]

print (final_z1_values)

However, we can combine this logic together into one array. Numpy supplies a `logical_and` function for this, but the simplest way to construct a single logic array satisfying two conditions is:

In [ ]:
# WARNING: Note the brackets here! `&` is evaluated before `>` so make sure you don't forget them!
y1_and_x1_logic = (pd_dataframe['y1'] > 0.5) & (reduced_x1 < 10)
final_z1_values = pd_dataframe['z1'][y1_and_x1_logic]
print (final_z1_values)

**EXERCISES**

* Find all values of `aa` for which `abs(x1 - 50)` is bigger than 20, `arccos(y1) > 45 degrees` and `z1**0.5 > 4`

## Methods of numpy arrays.

The numpy array class defines a number of `methods` for easily accessing information about the numpy array. Let's explore a few of these methods now.

### The `.any` and `.all` method

The `any` method can be used to check if at least one element in a boolean array is True. For example if we want to go back to our `some_numbers` array, and ask if it contains any number greater than 20 we can do:

In [ ]:
some_numbers = np.array([1,10,2,20,3,30,4,40,5,50,6,60])

if (some_numbers > 20).any():
    print ("There is at least one value in some_numbers larger than 20")

Similarly we can use the `all` method to ask if all numbers in `some_numbers` are larger than 20:

In [ ]:
some_numbers = np.array([1,10,2,20,3,30,4,40,5,50,6,60])

if (some_numbers > 20).all():
    print ("All numbers in some_numbers are larger than 20")
else:
    print ("Not all the numbers in some_numbers are larger than 20")

In both cases, these methods can also run on non boolean arrays:

In [ ]:
print (some_numbers.all())

In such cases it will be checking the "truth value" of each entry in the array. 0, or 0.0 or '' are False for integers, floats and strings respectively, anything else is True. This is the same as running `bool(value)` on all values in the array and then checking if any, or all, are True.

Let's take our pandas example again before setting some exercises:

In [ ]:
import numpy as np
import pandas as pd
def func_makedata(a):
    x1 = a**2
    y1 = np.cos(a)
    z1 = 3*a**2 
    return x1, y1, z1

aa = np.linspace(0.,10.,50)
x1, y1, z1 = func_makedata(aa)

data_dict = {}
data_dict['aa'] = aa
data_dict['x1'] = x1
data_dict['y1'] = y1
data_dict['z1'] = z1
pd_dataframe = pd.DataFrame(data_dict)
pd_dataframe

**EXERCISES**

* Are there any values in $\mathrm{y1}$ for which $\cos^{-1}(y1) > 3.0$ (radians)
* Do all of the values of $(x1 - 50) * (z1 - 150)$ have a magnitude smaller than 5000?

### The `np.where` function

Consider again our list of `some_numbers`. Let's say that we have a problem that asks: For your list of numbers create a new list storing $x^2$ if $x$ is smaller than 10 or $x^0.5$ is x is larger than or equal to 10. This is a perfect example for numpy's `where` function:

In [ ]:
# Let's let numpy explain what this function does first
np.where?

In [ ]:
some_numbers = np.array([1,10,2,20,3,30,4,40,5,50,6,60])
np.where(some_numbers >= 10, some_numbers**0.5, some_numbers**2)

Alternatively, if we just run this with the first argument:

In [ ]:
np.where(some_numbers >= 10)[0]

It returns the indexes of the points where some_numbers is larger than or equal to 10. This can also be used to index, so we can do:

In [ ]:
# Returns the points at positions 1, 3, 5, 7, 9 and 11
some_numbers[np.where(some_numbers >= 10)[0]]

This is equivalent to having done:

In [ ]:
some_numbers[some_numbers >= 10]

But sometimes it's nice to see the indexes.

**EXERCISE**

Create an array out of some_numbers from the following condition.

* If the number is larger than 20, store $x^3$. If the number is smaller than 20, store 1.

Then try the following:

* If the number is larger than 20, store $x^3$. If the number is larger than 4 but smaller than or equal to 20 store $x$, otherwise store 1.

## Transposing and reshaping arrays

Let's move on now to consider some multi-dimension arrays. Let's first consider the following 5 x 5 array of random numbers uniformly distributed between 0 and 10:

In [ ]:
nums = np.random.random_sample(size=[12,10]) * 10
print (nums)

To transpose this array is quite simple, numpy supplies a `.T` property:

In [ ]:
print (nums.T)

If we want to reshape the array we can use numpy's `reshape` function. For example if I want to "flatten" this array (note there is also a `flatten` function, but `reshape` is more general), we could do:

In [ ]:
np.reshape(nums, (120,))

This indicates that this should become a 1D array with 120 entries. We could also make this a 4 x 30 array or a 30 x 4 array with something like:

In [ ]:
np.reshape(nums, (4,30))

In [ ]:
np.reshape(nums, (30,4))

In all cases the order of the data is preserved. There is a bit more information on reshaping an array here:

https://www.w3resource.com/numpy/manipulation/reshape.php

If you try to do something invalid, the code will fail:

In [ ]:
np.reshape(nums, (30,10))

**EXERCISE**

Create a 2D array (x,y) whose values should be equal to $x**2 + y$ (so the value at `[0,1]` would be 1, the value at `[2,4]` would be $2^2 + 4 = 8$ and so on). The size of this array should be 42 x 20.

Take the transpose of this array. Now if this new array has coordinates (x1,y1) what is the value at x1 = 4, y1=5 going to be?

There are a number of different 2D shapes that you can reshape this array into (I counted 32). How many valid different sizes can you reshape this array into.

Reshape this array into a 5D shape, where no dimension has a size equal to 1.

## Matrices

Finally numpy defines a matrix type:

In [ ]:
nums = np.random.random_sample(size=[12,10]) * 10
nums = np.matrix(nums)

print (nums)

The matrix class is just a subclass of the array, but adds some additional functionality that can be useful for matrices. For example we can do matrix multiplication:

In [ ]:
nums1 = np.random.random_sample(size=[12,10]) * 10
nums2 = np.random.random_sample(size=[12,10]) * 10

prod = nums1 * nums2
# To multiply two numpy arrays they must be the same shape, and then every element is multiplied by its corresponding
# element in the other array. That's not how matrix multiplication works
print (prod.shape)
print()

nums1 = np.matrix(nums1)
nums2 = np.matrix(nums2)

# This isn't possible! I cannot multiply a 12x10 matrix by a 12x10 matrix!
# Let's catch this error and continue. Python's try/except can be used for this
# https://docs.python.org/3/tutorial/errors.html

try:
    nums1 * nums2
except ValueError as err:
    print ("This doesn't work! The code says:")
    print (err)

# But I can multiply a 12x10 matrix by a 10x12 matrix (the resulting matrix should be 12 x 12.)
# OR I can multiply a 10x12 matrix by a 12x10 matrix (resulting in a 10x10 matrix)

prod1 = nums1 * nums2.T
prod2 = nums1.T * nums2

print (prod1.shape, prod2.shape)

We can also compute the eigenvalues and eigenvectors of a square matrix (the following will work fine on either a matrix or 2D array object):

In [ ]:
# Make a diagonal matrix
nums1 = np.zeros([3,3])
nums1[0,0] = 1.
nums1[1,1] = 2.
nums1[2,2] = 3.
print(nums1)
# Or equivalently
print (np.diag([1.,2.,3.]))

evals, evecs = np.linalg.eig(nums1)

print (evals)
print (evecs)

**EXERCISE**

Generate a 2D 10x10 matrix of random numbers. Compute the eigenvectors and eigenvalues of this matrix.

Now try again but this time generate a symmetric matrix (where the value at `[x,y]` is equal to the value at `[y,x]`). What do you notice is different? Why is this?

In [1]:
import numpy as np
def func_makedata(a):
    x1 = a**2
    y1 = np.cos(a)
    z1 = 3*a**2 
    return x1, y1, z1

aa = np.linspace(0.,10.,50)
x1, y1, z1 = func_makedata(aa)
 

We now have four arrays (aa, x1, y1 and z1) which we might want to store. Let's start with storing a single array to disk.

In [2]:
np.savetxt('data_array_aa.dat', aa)

This stores the contents of the array "aa" to the file `data_array_aa.dat`. You can open this file in the SciServer window, check what is in it.

However, we might want to store all 4 of the arrays, to do this we need to *arrange* the data in such a way that this will work nicely. We can do

In [3]:
np.savetxt('data_array_aa2.dat', np.array([aa, x1, y1, z1]))

Again, you can see this in the SciServer window. In this case we have 4 rows, each containing N columns (where N is the length of aa and the other arrays). If we want to swap this to 4 columns and N rows we can do:

In [4]:
np.savetxt('data_array_aa3.dat', np.array([aa, x1, y1, z1]).T)

which will transpose the array and store it nicely.

In all these cases the operation to read the corresponding array back from disk is

In [5]:
file_data = np.loadtxt('data_array_aa3.dat')

`file_data` now stores the arrays we read from disk. To recover the y1 array (which was stored in the 3rd column) we would need to use *slicing* to recover this array as demonstrated previously:

In [6]:
y1_from_file = file_data[:,2]
print(y1 - y1_from_file)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0.]


An alternative to the `savetxt` command is the `savez` command. The advantages to `savez` are:
 * You can easily save and read back in numerous arrays
 * The arrays do not have to do the same length (or need to be shaped into a N x M 2D format)
 * It's much faster to read/save data using this command.

But the one major disadvantage is:

 * These files are not human readable, are not easily readable by anything other than numpy and may not be the same on machines with different CPU architecture (e.g. a file generated on a windows machine (x86_64) may not be readable on a raspberry pi (ARM))

Here's how this works with our datasets above

In [7]:
np.savez('data_array_aa.npz', aa=aa, x1=x1, y1=y1, z1=z1)

Try reading this file in the SciServer window!

To read this file back in with numpy you use:

In [8]:
file_data = np.load('data_array_aa.npz')
print(list(file_data))

['aa', 'x1', 'y1', 'z1']


`file_data` is implemented basically as a sub-class of python's dictionary. To access each of the arrays we do so as we would access entries in a python dictionary

In [9]:
y1_from_file = file_data['y1']
print(y1_from_file - y1)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0.]


**EXERCISE** Let's practice using this functionality. Write a function to do the following:
 * Create an array containing 1000 numbers uniformly distributed between 0 and $\pi$. Both 0 and $\pi$ should be in the array as the 1st and 1000th entries.
 * Create a second array storing $\sin(x)$
 * Create a third array storing $\cos^2(x) +1$
 * Create a fourth array storing $\mathrm{cosech}(x)$
 
Then write these 4 arrays to a file, read them back in, and check that you can recover the original arrays. Try using the different options illustrated above.

In [ ]:
# Exercise solution goes here


## Reading/Writing data to file with pandas

An alternative to numpy is the pandas module. pandas is very nice for reading/writing and manipulating tables of data. Let's illustrate some of the basic functionality here, first let's generate our data arrays again:


In [10]:
import numpy as np
import pandas as pd
def func_makedata(a):
    x1 = a**2
    y1 = np.cos(a)
    z1 = 3*a**2 
    return x1, y1, z1

aa = np.linspace(0.,10.,50)
x1, y1, z1 = func_makedata(aa)

Then we create a pandas `DataFrame` object to store our data. This is initialized from a dictionary, so we first put our data into a dictionary and then initialize the `DataFrame` object.

In [11]:
data_dict = {}
data_dict['aa'] = aa
data_dict['x1'] = x1
data_dict['y1'] = y1
data_dict['z1'] = z1
pd_dataframe = pd.DataFrame(data_dict)

Note that if we print the dataframe it looks much nicer than numpy arrays! (For this to look nice don't use the `print` function, as this is stuff integrated with Jupyter to make it look nice in the notebook)

In [12]:
pd_dataframe

aa          x1        y1          z1
0    0.000000    0.000000  1.000000    0.000000
1    0.204082    0.041649  0.979248    0.124948
2    0.408163    0.166597  0.917851    0.499792
3    0.612245    0.374844  0.818360    1.124531
4    0.816327    0.666389  0.684902    1.999167
5    1.020408    1.041233  0.523018    3.123698
6    1.224490    1.499375  0.339426    4.498126
7    1.428571    2.040816  0.141746    6.122449
8    1.632653    2.665556 -0.061817    7.996668
9    1.836735    3.373594 -0.262815   10.120783
10   2.040816    4.164931 -0.452904   12.494794
11   2.244898    5.039567 -0.624196   15.118701
12   2.448980    5.997501 -0.769580   17.992503
13   2.653061    7.038734 -0.883023   21.116202
14   2.857143    8.163265 -0.959816   24.489796
15   3.061224    9.371095 -0.996772   28.113286
16   3.265306   10.662224 -0.992357   31.986672
17   3.469388   12.036651 -0.946755   36.109954
18   3.673469   13.494377 -0.861857   40.483132
19   3.877551   15.035402 -0.741188   45.106206
20   4.081633   16.659725 -0.589756   49.979175
21   4.285714   18.367347 -0.413846   55.102041
22   4.489796   20.158267 -0.220759   60.474802
23   4.693878   22.032486 -0.018510   66.097459
24   4.897959   23.990004  0.184507   71.970012
25   5.102041   26.030820  0.379866   78.092461
26   5.306122   28.154935  0.559459   84.464806
27   5.510204   30.362349  0.715832   91.087047
28   5.714286   32.653061  0.842494   97.959184
29   5.918367   35.027072  0.934189  105.081216
30   6.122449   37.484382  0.987110  112.453145
31   6.326531   40.024990  0.999061  120.074969
32   6.530612   42.648896  0.969546  127.946689
33   6.734694   45.356102  0.899790  136.068305
34   6.938776   48.146606  0.792688  144.439817
35   7.142857   51.020408  0.652686  153.061224
36   7.346939   53.977509  0.485594  161.932528
37   7.551020   57.017909  0.298348  171.053728
38   7.755102   60.141608  0.098719  180.424823
39   7.959184   63.348605 -0.105008  190.045814
40   8.163265   66.638900 -0.304376  199.916701
41   8.367347   70.012495 -0.491112  210.037484
42   8.571429   73.469388 -0.657463  220.408163
43   8.775510   77.009579 -0.796527  231.028738
44   8.979592   80.633070 -0.902531  241.899209
45   9.183673   84.339858 -0.971075  253.019575
46   9.387755   88.129946 -0.999315  264.389838
47   9.591837   92.003332 -0.986078  276.009996
48   9.795918   95.960017 -0.931914  287.880050
49  10.000000  100.000000 -0.839072  300.000000

We can save this to file using built in methods to this `DataFrame` class. There's a few options here, but let's just show two `to_csv` which writes a human readable file, and `to_hdf` which writes an encoded file (but in a standard that is much more portable than numpy's binary files):

In [13]:
pd_dataframe.to_csv('data_array_pandas.csv')
# to_hdf can be used to store *multiple* DataFrames in a single file!
pd_dataframe.to_hdf('data_array_pandas.hdf', key='mydata')

To read this back in we can use pandas' `read_csv` and `read_hdf` functions:

In [14]:
# NOTE: As to_csv adds an index column in the output file, we have to not use this as a data column when reading
# the file back in. So we set index_col=0. See what happens when this argument is removed.
data_from_csv = pd.read_csv('data_array_pandas.csv', index_col=0)
data_from_hdf = pd.read_hdf('data_array_pandas.hdf', key='mydata')
data_from_csv

aa          x1        y1          z1
0    0.000000    0.000000  1.000000    0.000000
1    0.204082    0.041649  0.979248    0.124948
2    0.408163    0.166597  0.917851    0.499792
3    0.612245    0.374844  0.818360    1.124531
4    0.816327    0.666389  0.684902    1.999167
5    1.020408    1.041233  0.523018    3.123698
6    1.224490    1.499375  0.339426    4.498126
7    1.428571    2.040816  0.141746    6.122449
8    1.632653    2.665556 -0.061817    7.996668
9    1.836735    3.373594 -0.262815   10.120783
10   2.040816    4.164931 -0.452904   12.494794
11   2.244898    5.039567 -0.624196   15.118701
12   2.448980    5.997501 -0.769580   17.992503
13   2.653061    7.038734 -0.883023   21.116202
14   2.857143    8.163265 -0.959816   24.489796
15   3.061224    9.371095 -0.996772   28.113286
16   3.265306   10.662224 -0.992357   31.986672
17   3.469388   12.036651 -0.946755   36.109954
18   3.673469   13.494377 -0.861857   40.483132
19   3.877551   15.035402 -0.741188   45.106206
20   4.081633   16.659725 -0.589756   49.979175
21   4.285714   18.367347 -0.413846   55.102041
22   4.489796   20.158267 -0.220759   60.474802
23   4.693878   22.032486 -0.018510   66.097459
24   4.897959   23.990004  0.184507   71.970012
25   5.102041   26.030820  0.379866   78.092461
26   5.306122   28.154935  0.559459   84.464806
27   5.510204   30.362349  0.715832   91.087047
28   5.714286   32.653061  0.842494   97.959184
29   5.918367   35.027072  0.934189  105.081216
30   6.122449   37.484382  0.987110  112.453145
31   6.326531   40.024990  0.999061  120.074969
32   6.530612   42.648896  0.969546  127.946689
33   6.734694   45.356102  0.899790  136.068305
34   6.938776   48.146606  0.792688  144.439817
35   7.142857   51.020408  0.652686  153.061224
36   7.346939   53.977509  0.485594  161.932528
37   7.551020   57.017909  0.298348  171.053728
38   7.755102   60.141608  0.098719  180.424823
39   7.959184   63.348605 -0.105008  190.045814
40   8.163265   66.638900 -0.304376  199.916701
41   8.367347   70.012495 -0.491112  210.037484
42   8.571429   73.469388 -0.657463  220.408163
43   8.775510   77.009579 -0.796527  231.028738
44   8.979592   80.633070 -0.902531  241.899209
45   9.183673   84.339858 -0.971075  253.019575
46   9.387755   88.129946 -0.999315  264.389838
47   9.591837   92.003332 -0.986078  276.009996
48   9.795918   95.960017 -0.931914  287.880050
49  10.000000  100.000000 -0.839072  300.000000

**EXERCISE** Let's repeat the exercise above but now use pandas to read/write the arrays: Write a function to do the following:
 * Create an array containing 1000 numbers uniformly distributed between 0 and $\pi$. Both 0 and $\pi$ should be in the array as the 1st and 1000th entries.
 * Create a second array storing $\sin(x)$
 * Create a third array storing $\cos^2(x) +1$
 * Create a fourth array storing $\mathrm{cosech}(x)$
 
Then write these 4 arrays to a file, read them back in, and check that you can recover the original arrays. Use the different pandas options illustrated above.

**EXERCISE**

In this exercise, we'll be calculating the pressue of gas inside a closed 2D box as the box is heated. We'll be making use of the ideal gas law for this exercise

As a reminder, the ideal gas law is PV = NkT.

Here, P is the pressre, V is the volume, N is the number of molecules, k is the Boltzman constant, and T is the temperature. 

For the initial condition, we'll assume standard temperature i.e. 273.15 K. We'll assume the box is cubic with sides of 1 m length and contains 2.5 x 10^25 molecules of gas.

The Boltzman constant is 1.380e-23 J/K
1 atmosphere = 101325 N/m^2

1.  Write a function to calculate the pressure inside the box as the temperature is increased. N, T, and V should be inputs to this function. The pressure should be output in units of N/m^2

2. Verify that at the initial temperature (273.15 K) the pressure is 94236.75 N/m^2

3.  Compute how the pressure in the box varies as the temperature increases by 500K from the initial temperature (so going for temperatures between T = 273.15 K and T = 773.15 K). Your input array should contain 500 values. What is the  pressure in the box at the start and finish temperatures in units of atmospheres?

4. Make a labelled plot showing the input temperature and output pressure, in units of atmosphere. 

5. Save the output from this function as a table. 

6. Read in the saved table from the previous part of the exercise.

7. For each temperature in your table, calculate the average kinetic energy per molecule:

$$ \frac{1}{2}m v^{2} = \frac{3}{2}k_{B}T$$

8. Add this as a new column to your table, and save the result.

9. Make a labelled plot showing average kinetic energy as a function of pressure.



## SUMMARY

We've explored some additional features of numpy and pandas in this lecture. However, numpy (and scipy) have *many* features. Knowing all of them:

https://docs.scipy.org/doc/numpy/genindex.html

is not realistic! So often when approaching a problem it's worth formulating what you want to do and then checking if a solution already exists for this problem. Clearly formulating the problem maximizes the chance of finding a useful stackoverflow post!